beacon-chain/rpc/eth/validator/handlers.go

```diff
@@ -98,14 +78,110 @@ func (s *Server) GetAggregateAttestation(w http.ResponseWriter, r *http.Request)
+				BeaconBlockRoot: hexutil.Encode(match.GetData().BeaconBlockRoot),
+				Source: &structs.Checkpoint{
+					Epoch: strconv.FormatUint(uint64(match.GetData().Source.Epoch), 10),
+					Root:  hexutil.Encode(match.GetData().Source.Root),
+				},
+				Target: &structs.Checkpoint{
+					Epoch: strconv.FormatUint(uint64(match.GetData().Target.Epoch), 10),
+					Root:  hexutil.Encode(match.GetData().Target.Root),
+				},
+			},
+			Signature:     hexutil.Encode(match.GetSignature()),
+			CommitteeBits: hexutil.Encode(match.CommitteeBitsVal().Bytes()),
+		}
+	} else {
+		resp.Data = &structs.Attestation{
+			AggregationBits: hexutil.Encode(match.GetAggregationBits()),
+			Data: &structs.AttestationData{
+				Slot:            strconv.FormatUint(uint64(match.GetData().Slot), 10),
+				CommitteeIndex:  strconv.FormatUint(uint64(match.GetData().CommitteeIndex), 10),
+				BeaconBlockRoot: hexutil.Encode(match.GetData().BeaconBlockRoot),
+				Source: &structs.Checkpoint{
+					Epoch: strconv.FormatUint(uint64(match.GetData().Source.Epoch), 10),
+					Root:  hexutil.Encode(match.GetData().Source.Root),
+				},
+				Target: &structs.Checkpoint{
+					Epoch: strconv.FormatUint(uint64(match.GetData().Target.Epoch), 10),
+					Root:  hexutil.Encode(match.GetData().Target.Root),
+				},
+			},
+			Signature: hexutil.Encode(match.GetSignature()),
+		}
+	}
+	httputil.WriteJson(w, resp)
+}
+
+func (s *Server) aggregateAttestation(w http.ResponseWriter, slot, index uint64, attDataRoot []byte) ethpbalpha.Att {
+	var match ethpbalpha.Att
+	var err error
+
+	match, err = matchingAtt(s.AttestationsPool.AggregatedAttestations(), primitives.Slot(slot), attDataRoot, primitives.CommitteeIndex(index))
+	if err != nil {
+		httputil.HandleError(w, "Could not get matching attestation: "+err.Error(), http.StatusInternalServerError)
+		return nil
+	}
+	if match == nil {
+		atts, err := s.AttestationsPool.UnaggregatedAttestations()
+		if err != nil {
+			httputil.HandleError(w, "Could not get unaggregated attestations: "+err.Error(), http.StatusInternalServerError)
+			return nil
+		}
+		match, err = matchingAtt(atts, primitives.Slot(slot), attDataRoot, primitives.CommitteeIndex(index))
+		if err != nil {
+			httputil.HandleError(w, "Could not get matching attestation: "+err.Error(), http.StatusInternalServerError)
+			return nil
+		}
+	}
+	if match == nil {
+		httputil.HandleError(w, "No matching attestation found", http.StatusNotFound)
+		return nil
+	}
+	return match
+}
+
+func matchingAtt(atts []ethpbalpha.Att, slot primitives.Slot, attDataRoot []byte, index primitives.CommitteeIndex) (ethpbalpha.Att, error) {
 	for _, att := range atts {
 		if att.GetData().Slot == slot {
 			root, err := att.GetData().HashTreeRoot()
 			if err != nil {
 				return nil, errors.Wrap(err, "could not get attestation data root")
 			}
-			if bytes.Equal(root[:], attDataRoot) {
+			if bytes.Equal(root[:], attDataRoot) && att.GetData().CommitteeIndex == index {
```

Comment:
wonder if it's possible to have an old attestation that doesn't have a committeeIndex and have it match with 0

Reply:
This code is incorrect. We should ignore the committee index entirely for pre-Electra attestations because the attestation data has the committee index, meaning that hashing the data is enough to get a unique attestation. This is not true post-Electra because `GetData().CommitteeIndex` will always be `0`, meaning that two attestations for different committees may have the same attestation data root. We need to check `CommitteeBitsVal().BitAt(index)`.